# Tarea 1

Primero importamos las librerías necesarias

In [1]:
from pylab import*
import numpy as np
import pandas as pd
import sqldf
import scipy.stats
from scipy.special import gamma
import seaborn as sns
from scipy.stats import spearmanr
from scipy.stats import beta
from sklearn import linear_model

# Pregunta 1

In [2]:
path = "C:\\Users\\Arath Reyes\\Documents\\GitHub\\Seminario-ML-DS\\data\\"
bd_ini_int = pd.read_csv(path + "base_ini_interna.csv")
bd_ini_pt1 = pd.read_csv(path + "base_ini_parte1.csv")
bd_comp3 = pd.read_csv(path + "base_comportamiento_3.csv")
# Archivos .txt
bd_comp2 = pd.read_csv(path + "base_comportamiento_2.txt", sep = " ")
bd_ini_ext = pd.read_csv(path+"base_ini_externa.txt",sep = " ")
bd_inver = pd.read_csv(path + "base_inversion.txt", sep = " ")

Empezamos el análisis exploratorio de las bases de datos.

### base_ini_interna

In [3]:
bd_ini_int = bd_ini_int.sort_values(by = ["fecha_inicio"])
bd_ini_int.head()

_id_ fecha_inicio  comp_interno1 comp_interno2  comp_interno3  \
17236  2333   2016-01-04    6079.027356          b'1'           30.0   
47365   488   2016-01-04     911.854103          b'0'           19.0   
21961  1964   2016-01-04       0.000000          b'1'           19.0   
49384  1220   2016-01-04    2482.269504          b'0'            9.0   
16983  1226   2016-01-04       0.000000          b'1'            3.0   

       comp_interno4  categoria1  categoria2  categoria3  
17236           50.0         2.0         9.0       175.0  
47365           48.0         3.0         3.0       164.0  
21961           53.0         4.0         3.0       164.0  
49384           70.0         3.0         9.0       154.0  
16983           75.0         3.0         9.0       148.0

In [4]:
bd_ini_int.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 50000 entries, 17236 to 32134
Data columns (total 9 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   _id_           50000 non-null  int64  
 1   fecha_inicio   50000 non-null  object 
 2   comp_interno1  50000 non-null  float64
 3   comp_interno2  50000 non-null  object 
 4   comp_interno3  50000 non-null  float64
 5   comp_interno4  50000 non-null  float64
 6   categoria1     50000 non-null  float64
 7   categoria2     50000 non-null  float64
 8   categoria3     50000 non-null  float64
dtypes: float64(6), int64(1), object(2)
memory usage: 3.8+ MB


In [5]:
bd_ini_int.nunique()

_id_             50000
fecha_inicio      1557
comp_interno1      680
comp_interno2        2
comp_interno3       32
comp_interno4       55
categoria1           6
categoria2          13
categoria3          32
dtype: int64

Notamos que no tenemos valores nulos en id . También, obtenemos el número de valores distintos que puede tener cada variable (columna) en la base de datos.

Como todos los id son distintos y tenemos en total el mismo número de id que el total de id que es de 50000, lo que implica que no hay id repetidos.

Por otro lado, las demás variables tienen un número menor a 50000 de valores distintos, es decir, hay valores que se repiten.

In [6]:
for col in bd_ini_int.columns[3:]:
    print("Valores únicos de "+col+": {0}\n".format(sorted(bd_ini_int[col].unique())))

Valores únicos de comp_interno2: ["b'0'", "b'1'"]

Valores únicos de comp_interno3: [1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0, 9.0, 10.0, 11.0, 12.0, 13.0, 14.0, 15.0, 16.0, 17.0, 18.0, 19.0, 20.0, 21.0, 22.0, 23.0, 24.0, 25.0, 26.0, 27.0, 28.0, 29.0, 30.0, 31.0, 32.0]

Valores únicos de comp_interno4: [24.0, 25.0, 27.0, 28.0, 29.0, 30.0, 31.0, 32.0, 33.0, 34.0, 35.0, 36.0, 37.0, 38.0, 39.0, 40.0, 41.0, 42.0, 43.0, 44.0, 45.0, 46.0, 47.0, 48.0, 49.0, 50.0, 51.0, 52.0, 53.0, 54.0, 55.0, 56.0, 57.0, 58.0, 59.0, 60.0, 61.0, 62.0, 63.0, 64.0, 65.0, 66.0, 67.0, 68.0, 69.0, 70.0, 71.0, 72.0, 73.0, 74.0, 75.0, 76.0, 77.0, 78.0, 79.0]

Valores únicos de categoria1: [1.0, 2.0, 3.0, 4.0, 5.0, 6.0]

Valores únicos de categoria2: [1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0, 9.0, 10.0, 11.0, 12.0, 13.0]

Valores únicos de categoria3: [146.0, 147.0, 148.0, 149.0, 150.0, 151.0, 152.0, 153.0, 154.0, 155.0, 156.0, 157.0, 158.0, 159.0, 160.0, 161.0, 162.0, 163.0, 164.0, 165.0, 166.0, 167.0, 168.0, 169.0, 170.

### base_ini_externa

In [7]:
bd_ini_ext.head()

_id_  comp_externo1  comp_externo2  comp_externo3  comp_externo4
0  78213     260.869565      25.854639           56.0         -230.0
1  76457      40.579710       2.872738           62.0         -230.0
2  40666            NaN            NaN          119.0         -230.0
3   8329     349.275362      34.616490           94.0         -230.0
4  68389     276.811594      27.434645          125.0         -230.0

In [8]:
bd_ini_ext.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 5 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   _id_           50000 non-null  int64  
 1   comp_externo1  47831 non-null  float64
 2   comp_externo2  44410 non-null  float64
 3   comp_externo3  49994 non-null  float64
 4   comp_externo4  50000 non-null  float64
dtypes: float64(4), int64(1)
memory usage: 1.9 MB


In [9]:
bd_ini_ext.nunique()

_id_             50000
comp_externo1     7533
comp_externo2     6982
comp_externo3      237
comp_externo4        6
dtype: int64

Del mismo modo, como en la base anterior, tenemos 50,000 id distintos que es igual al total de id, es decir, que no se repiten.

Por otro lado, los componentes externos 1,2 y 3 tienen valores nulos.

### base_ini_parte1

In [10]:
bd_ini_pt1.head()

_id_    _id2_  comp_mixto1
0  78213  1798899      0.33522
1  76457  1758511      0.33334
2  40666   935318      0.29186
3   8329   191567      0.29112
4  68389  1572947      0.27318

In [11]:
bd_ini_pt1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 3 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   _id_         50000 non-null  int64  
 1   _id2_        50000 non-null  int64  
 2   comp_mixto1  50000 non-null  float64
dtypes: float64(1), int64(2)
memory usage: 1.1 MB


In [12]:
bd_ini_pt1.nunique()

_id_           50000
_id2_          50000
comp_mixto1     7098
dtype: int64

En esta base tampoco hay entradas nulas y notamos que ambos id no están repetidos

### base_comportamiento_2

In [13]:
bd_comp2 = bd_comp2.sort_values(by = "mes_informacion")
bd_comp2.head()

num_caso      mes_informacion  Tiempo_aparicion
32240      1220  2016-01-01 00:00:00               0.0
44912      1704  2016-01-01 00:00:00               0.0
74499      2824  2016-01-01 00:00:00               0.0
20479       775  2016-01-01 00:00:00               0.0
27608      1049  2016-01-01 00:00:00               0.0

In [ ]:
bd_comp2[60:67]

In [ ]:
bd_comp2.tail(20)

Al ver los primeros y úlitmos datos de la base bd_comp2, notamos que por cada número de caso hay varios registros. Para cada número de caso hay un mes de información y tiempo de aparición. Un mismo caso tiene varias apariciones. También que los números de casos no son seriados, es decir, está resigtrado el caso 1 , pero no hay algo al respecto del caso 2 al 6, mientras que para el caso 7 si hay registros.

Lo mismo sucede en varias partes de la base de datos como se observa en los outputs anteriores. Esto significa que ciertos casos iniciaron un trpamite que otros casos no hicieron. Además, de este trámite se lleva un registro del mes y el tiempo de aparición, lo cual se explicará detalladamente más adelante.

In [14]:
bd_comp2.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1455536 entries, 32240 to 1455535
Data columns (total 3 columns):
 #   Column            Non-Null Count    Dtype  
---  ------            --------------    -----  
 0   num_caso          1455536 non-null  int64  
 1   mes_informacion   1455536 non-null  object 
 2   Tiempo_aparicion  1455536 non-null  float64
dtypes: float64(1), int64(1), object(1)
memory usage: 44.4+ MB


No tenemos valores nulos.

In [15]:
bd_comp2.nunique()

num_caso            49352
mes_informacion        65
Tiempo_aparicion       65
dtype: int64

In [ ]:
for col in bd_comp2.columns[1:]:
    print("Valores únicos de "+col+": {0}\n".format(sorted(bd_comp2[col].unique())))

Con lo anterior, corroboramos que el número de valores úncios de num_caso es de 49,352, el cual es menor a el total de registros de num_casos. Esto nos indica que hay num_caso repetidos.

Por otro lado, las variables mes_informacion y Tiempo_aparicion tienen el mismo número de valores únicos, el cual es de 65.

Además, se observa algo particular. Que para un mismo número de caso, este puede tener 64 apariciones, es de 64 registros. Los cuales están en orden cronológico , ási, mes_informacion es el mes en el que el caso hace aparicion y Tiempo_aparicion es número que se le asigna a dicha aparición dependiendo de su orden cronológico.

Para visualizarlo mejor, voy a acomodar la tabla por Tiempo_aparicion, esto, con la finalidad de identificar cuales numeros de casos llegan a tener 64 apariciones.

In [16]:
bd_comp2["Tiempo_aparicion"].unique()

array([ 0.,  1.,  2.,  3.,  4.,  5.,  6.,  7.,  8.,  9., 10., 11., 12.,
       13., 14., 15., 16., 17., 18., 19., 20., 21., 22., 23., 24., 25.,
       26., 27., 28., 29., 30., 31., 32., 33., 34., 35., 36., 37., 38.,
       39., 40., 41., 42., 43., 44., 45., 46., 47., 48., 49., 50., 51.,
       52., 53., 54., 55., 56., 57., 58., 59., 60., 61., 62., 63., 64.])

### base_ini_comportamiento3

In [17]:
bd_comp3 = bd_comp3.sort_values(by ="mes_informacion")
bd_comp3.head()

num_caso      mes_informacion  comportamiento_j
32240      1220  2016-01-01 00:00:00               0.0
44912      1704  2016-01-01 00:00:00               0.0
74499      2824  2016-01-01 00:00:00               0.0
20479       775  2016-01-01 00:00:00               0.0
27608      1049  2016-01-01 00:00:00               0.0

Para cada número de caso hay un mes de informacipoin y un comportamiento j en dicho mes de información. Además, se observa algo interesante. Al ver los primeros 65 datos de bd_comp3 y compararlos con los primeros 65 datos de bd_comp2, parece ser que los renglones correpsonden al mismo número de caos con el mismo mes de información. También observamos lo mismo de las colas de dichas bases de datos.

Por otro lado, el bd_comp3 parece ser que se salta los mismos numeros de casos que en bd_comp2, es decir, por ejemplo, en bd_comp2 no había registros correpsondientes a los casos 2,3,4,5,6 y parece ser que en bd_comp3 tampoco. Más adelante haremos unas consultas para corroborar las observaciones que se plantean en este párrafo acerca si tienen o no los mismos números de casos.

In [18]:
bd_comp3.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1455536 entries, 32240 to 1455535
Data columns (total 3 columns):
 #   Column            Non-Null Count    Dtype  
---  ------            --------------    -----  
 0   num_caso          1455536 non-null  int64  
 1   mes_informacion   1455536 non-null  object 
 2   comportamiento_j  1455536 non-null  float64
dtypes: float64(1), int64(1), object(1)
memory usage: 44.4+ MB


In [19]:
bd_comp3.nunique()

num_caso            49352
mes_informacion        65
comportamiento_j       18
dtype: int64

In [20]:
bd_comp3["comportamiento_j"].unique()

array([ 0.,  1.,  2.,  3.,  4.,  5.,  6.,  7.,  8.,  9., 10., 13., 11.,
       12., 16., 18., 14., 15.])

In [21]:
for col in bd_comp3.columns[1:]:
    print("Valores únicos de "+col+": {0}\n".format(sorted(bd_comp3[col].unique())))

Valores únicos de mes_informacion: ['2016-01-01 00:00:00', '2016-02-01 00:00:00', '2016-03-01 00:00:00', '2016-04-01 00:00:00', '2016-05-01 00:00:00', '2016-06-01 00:00:00', '2016-07-01 00:00:00', '2016-08-01 00:00:00', '2016-09-01 00:00:00', '2016-10-01 00:00:00', '2016-11-01 00:00:00', '2016-12-01 00:00:00', '2017-01-01 00:00:00', '2017-02-01 00:00:00', '2017-03-01 00:00:00', '2017-04-01 00:00:00', '2017-05-01 00:00:00', '2017-06-01 00:00:00', '2017-07-01 00:00:00', '2017-08-01 00:00:00', '2017-09-01 00:00:00', '2017-10-01 00:00:00', '2017-11-01 00:00:00', '2017-12-01 00:00:00', '2018-01-01 00:00:00', '2018-02-01 00:00:00', '2018-03-01 00:00:00', '2018-04-01 00:00:00', '2018-05-01 00:00:00', '2018-06-01 00:00:00', '2018-07-01 00:00:00', '2018-08-01 00:00:00', '2018-09-01 00:00:00', '2018-10-01 00:00:00', '2018-11-01 00:00:00', '2018-12-01 00:00:00', '2019-01-01 00:00:00', '2019-02-01 00:00:00', '2019-03-01 00:00:00', '2019-04-01 00:00:00', '2019-05-01 00:00:00', '2019-06-01 00:00:00'

Efectivamente, tenemos 65 valores únicos en mes_informacion, tal como en la base bd_comp2. Por otro lado, la variable comportamiento_j tiene 15 valores únicos. Así, de las bases bd_comp2 y bd_comp3 se infiere que dichas tablas toenen registro de un número de caso si este empieza un cierto trámite específico, pues no todos los numero de casos están registrados. Una vez empezado el trámite, se rastrea su comportamiento , tiempo de aparición y el mes de información.

A continuación haremos las consultas para verificar que bd_comp2 y bd_comp 3 tengan los mismos números de casos con el mismo número de registros

In [ ]:
#Para la base bd_comp2
query1 ="""
       select num_caso as _id_,count(num_caso) as cantidad_apariciones
       from bd_comp2 group by num_caso
       """
num_tramites_bd_comp2=sqldf.run(query1)
num_tramites_bd_comp2.head(15)

In [ ]:
#Para la base bd_comp3
query2 ="""
       select num_caso as _id_,count(num_caso) as cantidad_apariciones
       from bd_comp3 group by num_caso
       """
num_tramites_bd_comp3=sqldf.run(query2)
num_tramites_bd_comp3.head(15)

Efectivamente, parece ser que bd_comp2 y bd_comp3 tienen los mismo número de casos con el mismo número de registros cada uno. Lo que implica que dichas tablas se pueden unir. Adempas tmabipen observamos que tiene el mismo mes de información para cada caso.

Por último, vamos a acomodar bd_comp3 por mes_informacion, tal como lo hicimos con bd_comp2.

In [ ]:
#Acomodar bd_comp3 por mes_informacion
bd_comp3 = bd_comp3.sort_values(by ="mes_informacion")
bd_comp3.head()

### base_inversion

In [22]:
bd_inver.head()

_id2_  inversion
0  1758511   0.041039
1  1572947  77.422135
2  1051836  90.539953
3  1223255  90.205018
4  1894602   1.071354

In [23]:
bd_inver.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 17536 entries, 0 to 17535
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   _id2_      17536 non-null  int64  
 1   inversion  17536 non-null  float64
dtypes: float64(1), int64(1)
memory usage: 274.1 KB


In [24]:
bd_inver.nunique()

_id2_        17536
inversion    17385
dtype: int64

Así, de las observaciones anteriores, se puede hacer un merge con bd_ini_int y bd_ini_ext por medio de la variable id.

También, nos interesa unir bd_comp 2 con bd_comp3 por medio de la variable num_caso y mes_informacion.

Por último, se uniran las bases bd_inv con bd_ini_pt1 por medio de las variable id2

### Análisis y agrupamiento de las bases

In [25]:
bd_ini_merged = bd_ini_int.merge(bd_ini_ext, how = "left", on = "_id_")
bd_ini_merged.head()

_id_ fecha_inicio  comp_interno1 comp_interno2  comp_interno3  \
0  2333   2016-01-04    6079.027356          b'1'           30.0   
1   488   2016-01-04     911.854103          b'0'           19.0   
2  1964   2016-01-04       0.000000          b'1'           19.0   
3  1220   2016-01-04    2482.269504          b'0'            9.0   
4  1226   2016-01-04       0.000000          b'1'            3.0   

   comp_interno4  categoria1  categoria2  categoria3  comp_externo1  \
0           50.0         2.0         9.0       175.0     324.637681   
1           48.0         3.0         3.0       164.0      15.942029   
2           53.0         4.0         3.0       164.0     144.927536   
3           70.0         3.0         9.0       154.0     727.536232   
4           75.0         3.0         9.0       148.0      30.434783   

   comp_externo2  comp_externo3  comp_externo4  
0      32.174662           91.0         -230.0  
1            NaN          132.0         -230.0  
2      14.363689           77.0         -230.0  
3      72.105717           75.0         -230.0  
4       3.016375          121.0         -230.0

In [26]:
bd_ini_merged.nunique()

_id_             50000
fecha_inicio      1557
comp_interno1      680
comp_interno2        2
comp_interno3       32
comp_interno4       55
categoria1           6
categoria2          13
categoria3          32
comp_externo1     7533
comp_externo2     6982
comp_externo3      237
comp_externo4        6
dtype: int64

In [27]:
bd_ini_merged.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 50000 entries, 0 to 49999
Data columns (total 13 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   _id_           50000 non-null  int64  
 1   fecha_inicio   50000 non-null  object 
 2   comp_interno1  50000 non-null  float64
 3   comp_interno2  50000 non-null  object 
 4   comp_interno3  50000 non-null  float64
 5   comp_interno4  50000 non-null  float64
 6   categoria1     50000 non-null  float64
 7   categoria2     50000 non-null  float64
 8   categoria3     50000 non-null  float64
 9   comp_externo1  47831 non-null  float64
 10  comp_externo2  44410 non-null  float64
 11  comp_externo3  49994 non-null  float64
 12  comp_externo4  50000 non-null  float64
dtypes: float64(10), int64(1), object(2)
memory usage: 5.3+ MB


In [28]:
bd_comp_merged = bd_comp2.groupby(["num_caso"], as_index=False)["mes_informacion"].max()
bd_comp_merged.head()

num_caso      mes_informacion
0         1  2021-05-01 00:00:00
1         7  2018-05-01 00:00:00
2         8  2021-04-01 00:00:00
3         9  2017-04-01 00:00:00
4        10  2017-05-01 00:00:00

Hacemos el groupby con el máximo sobre mes_informacion, pues nos interesa la informacion más reciente de cada cliente.

In [29]:
bd_comp_merged = bd_comp_merged.merge(bd_comp2, how = "left", on = ["num_caso", "mes_informacion"])
bd_comp_merged.head()

num_caso      mes_informacion  Tiempo_aparicion
0         1  2021-05-01 00:00:00              61.0
1         7  2018-05-01 00:00:00              28.0
2         8  2021-04-01 00:00:00              51.0
3         9  2017-04-01 00:00:00              14.0
4        10  2017-05-01 00:00:00              15.0

In [30]:
bd_comp_merged.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 49352 entries, 0 to 49351
Data columns (total 3 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   num_caso          49352 non-null  int64  
 1   mes_informacion   49352 non-null  object 
 2   Tiempo_aparicion  49352 non-null  float64
dtypes: float64(1), int64(1), object(1)
memory usage: 1.5+ MB


In [31]:
bd_aux = bd_comp3.groupby(["num_caso"], as_index = False)["comportamiento_j"].mean()
bd_aux.head()

num_caso  comportamiento_j
0         1          0.112903
1         7          0.827586
2         8          0.063492
3         9          0.000000
4        10          0.062500

In [32]:
bd_aux = bd_aux.sort_values(by ="num_caso")
bd_comp_merged = bd_comp_merged.sort_values(by = "num_caso")

In [33]:
bd_aux.head()

num_caso  comportamiento_j
0         1          0.112903
1         7          0.827586
2         8          0.063492
3         9          0.000000
4        10          0.062500

In [34]:
bd_comp_merged.head()

num_caso      mes_informacion  Tiempo_aparicion
0         1  2021-05-01 00:00:00              61.0
1         7  2018-05-01 00:00:00              28.0
2         8  2021-04-01 00:00:00              51.0
3         9  2017-04-01 00:00:00              14.0
4        10  2017-05-01 00:00:00              15.0

In [35]:
bd_comp_merged["num_caso"].equals( bd_aux["num_caso"])

True

In [36]:
bd_comp_merged["mean_comportamiento_j"] = bd_aux["comportamiento_j"]

In [37]:
bd_comp_merged.head()

num_caso      mes_informacion  Tiempo_aparicion  mean_comportamiento_j
0         1  2021-05-01 00:00:00              61.0               0.112903
1         7  2018-05-01 00:00:00              28.0               0.827586
2         8  2021-04-01 00:00:00              51.0               0.063492
3         9  2017-04-01 00:00:00              14.0               0.000000
4        10  2017-05-01 00:00:00              15.0               0.062500

In [38]:
bd_comp_merged.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 49352 entries, 0 to 49351
Data columns (total 4 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   num_caso               49352 non-null  int64  
 1   mes_informacion        49352 non-null  object 
 2   Tiempo_aparicion       49352 non-null  float64
 3   mean_comportamiento_j  49352 non-null  float64
dtypes: float64(2), int64(1), object(1)
memory usage: 1.9+ MB


In [39]:
bd_inv_merged = bd_ini_pt1.merge(bd_inver, how = "left", on = "_id2_")
bd_inv_merged.head()

_id_    _id2_  comp_mixto1  inversion
0  78213  1798899      0.33522        NaN
1  76457  1758511      0.33334   0.041039
2  40666   935318      0.29186        NaN
3   8329   191567      0.29112        NaN
4  68389  1572947      0.27318  77.422135

In [40]:
bd_inv_merged.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 50000 entries, 0 to 49999
Data columns (total 4 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   _id_         50000 non-null  int64  
 1   _id2_        50000 non-null  int64  
 2   comp_mixto1  50000 non-null  float64
 3   inversion    17536 non-null  float64
dtypes: float64(2), int64(2)
memory usage: 1.9 MB


In [41]:
bd_inv_merged.nunique()

_id_           50000
_id2_          50000
comp_mixto1     7098
inversion      17385
dtype: int64

### Veamos en dónde y cuales id's coinciden según la base.

In [42]:
x = set(bd_ini_pt1["_id_"])
print(len(x))
y = set(bd_ini_pt1["_id2_"])
print(len(y))
z = set(bd_comp2["num_caso"])
print(len(z))
w = set(bd_comp3["num_caso"])
print(len(w))
u = set(bd_ini_int["_id_"])
print(len(u))
v = set(bd_ini_ext["_id_"])
print(len(v))

50000
50000
49352
49352
50000
50000


In [43]:
x_y = y.intersection(x)
print(len(x_y))
x_z = x.intersection(z)
print(len(x_z))
x_w = x.intersection(w)
print(len(x_w))
y_z = y.intersection(z)
print(len(y_z))
y_w = y.intersection(w)
print(len(y_w))
z_w = z.intersection(w)
print(len(z_w))
u_v = u.intersection(v)
print(len(u_v))

1069
49352
49352
1053
1053
49352
50000


De lo siguiente podemos observar que, en gran medida, la columna _id_ de bd_ini_pt1 coincide con num_caso de ambas bases de comportamiento, por lo que cruce de esas bases con bd_ini_pt1 debería ser con la columna _id_. De la misma manera, podemos observar que la columna _id_ coincide en las bases externa e interna. Veamos si estas mismas coincide con el _id_ de bd_ini_pt1.

In [44]:
u_v_x = u_v.intersection(x)
len(u_v_x)

50000

Efectivamente, estos id's coinciden, por lo que el agrupamiento de estas bases debería ser sobre la columna id.

In [45]:
bd_id = bd_ini_pt1.merge(bd_ini_int, how = "left", on = "_id_")
bd_id = bd_id.sort_values(by = "fecha_inicio")
bd_id.head()

_id_  _id2_  comp_mixto1 fecha_inicio  comp_interno1 comp_interno2  \
17236  2333  53659      0.32856   2016-01-04    6079.027356          b'1'   
47365   488  11224      0.38498   2016-01-04     911.854103          b'0'   
21961  1964  45172      0.34780   2016-01-04       0.000000          b'1'   
49384  1220  28060      0.28422   2016-01-04    2482.269504          b'0'   
16983  1226  28198      0.31500   2016-01-04       0.000000          b'1'   

       comp_interno3  comp_interno4  categoria1  categoria2  categoria3  
17236           30.0           50.0         2.0         9.0       175.0  
47365           19.0           48.0         3.0         3.0       164.0  
21961           19.0           53.0         4.0         3.0       164.0  
49384            9.0           70.0         3.0         9.0       154.0  
16983            3.0           75.0         3.0         9.0       148.0

In [46]:
bd_id.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 50000 entries, 17236 to 32134
Data columns (total 11 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   _id_           50000 non-null  int64  
 1   _id2_          50000 non-null  int64  
 2   comp_mixto1    50000 non-null  float64
 3   fecha_inicio   50000 non-null  object 
 4   comp_interno1  50000 non-null  float64
 5   comp_interno2  50000 non-null  object 
 6   comp_interno3  50000 non-null  float64
 7   comp_interno4  50000 non-null  float64
 8   categoria1     50000 non-null  float64
 9   categoria2     50000 non-null  float64
 10  categoria3     50000 non-null  float64
dtypes: float64(7), int64(2), object(2)
memory usage: 4.6+ MB


In [47]:
bd_id.nunique()

_id_             50000
_id2_            50000
comp_mixto1       7098
fecha_inicio      1557
comp_interno1      680
comp_interno2        2
comp_interno3       32
comp_interno4       55
categoria1           6
categoria2          13
categoria3          32
dtype: int64

In [48]:
bd_final = bd_id.merge(bd_inv_merged, how = "inner", on = ["_id2_", "_id_", "comp_mixto1"])
bd_final.head()

_id_  _id2_  comp_mixto1 fecha_inicio  comp_interno1 comp_interno2  \
0  2333  53659      0.32856   2016-01-04    6079.027356          b'1'   
1   488  11224      0.38498   2016-01-04     911.854103          b'0'   
2  1964  45172      0.34780   2016-01-04       0.000000          b'1'   
3  1220  28060      0.28422   2016-01-04    2482.269504          b'0'   
4  1226  28198      0.31500   2016-01-04       0.000000          b'1'   

   comp_interno3  comp_interno4  categoria1  categoria2  categoria3  inversion  
0           30.0           50.0         2.0         9.0       175.0        NaN  
1           19.0           48.0         3.0         3.0       164.0        NaN  
2           19.0           53.0         4.0         3.0       164.0        NaN  
3            9.0           70.0         3.0         9.0       154.0        NaN  
4            3.0           75.0         3.0         9.0       148.0   0.311552

In [49]:
bd_final.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 50000 entries, 0 to 49999
Data columns (total 12 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   _id_           50000 non-null  int64  
 1   _id2_          50000 non-null  int64  
 2   comp_mixto1    50000 non-null  float64
 3   fecha_inicio   50000 non-null  object 
 4   comp_interno1  50000 non-null  float64
 5   comp_interno2  50000 non-null  object 
 6   comp_interno3  50000 non-null  float64
 7   comp_interno4  50000 non-null  float64
 8   categoria1     50000 non-null  float64
 9   categoria2     50000 non-null  float64
 10  categoria3     50000 non-null  float64
 11  inversion      17536 non-null  float64
dtypes: float64(8), int64(2), object(2)
memory usage: 5.0+ MB


In [50]:
bd_final.sort_values(by = "_id_")

_id_    _id2_  comp_mixto1 fecha_inicio  comp_interno1 comp_interno2  \
1864       1       23      0.38000   2016-03-31    3039.513678          b'1'   
238        7      161      0.33360   2016-01-19    1547.112462          b'1'   
718        8      184      0.27898   2016-02-10    1215.805471          b'1'   
909        9      207      0.38000   2016-02-18       0.000000          b'1'   
914       10      230      0.37050   2016-02-18       0.000000          b'1'   
...      ...      ...          ...          ...            ...           ...   
49735  98987  2276701      0.33712   2021-04-13       0.000000          b'1'   
49764  98988  2276724      0.26010   2021-04-14    2026.342452          b'1'   
49996  98990  2276770      0.30434   2021-04-29       0.000000          b'1'   
49992  98992  2276816      0.30506   2021-04-29       0.000000          b'1'   
49904  98997  2276931      0.35322   2021-04-21       0.000000          b'0'   

       comp_interno3  comp_interno4  categoria1  categoria2  categoria3  \
1864            26.0           67.0         3.0         3.0       171.0   
238             28.0           75.0         1.0        11.0       173.0   
718             14.0           55.0         2.0         6.0       159.0   
909             11.0           55.0         2.0         9.0       156.0   
914             11.0           56.0         2.0         6.0       156.0   
...              ...            ...         ...         ...         ...   
49735           23.0           51.0         3.0         9.0       168.0   
49764           19.0           63.0         3.0         9.0       164.0   
49996           17.0           60.0         2.0         9.0       162.0   
49992           15.0           47.0         1.0         9.0       160.0   
49904           15.0           43.0         1.0         9.0       160.0   

         inversion  
1864     52.500490  
238       0.047480  
718     227.537211  
909       0.921995  
914      39.259568  
...            ...  
49735          NaN  
49764  2596.865519  
49996   107.495449  
49992     9.822964  
49904          NaN  

[50000 rows x 12 columns]

De esta manera sólo nos falta agrupar bd_comp_merged y bd_final

In [51]:
bd_comp_merged.head()

num_caso      mes_informacion  Tiempo_aparicion  mean_comportamiento_j
0         1  2021-05-01 00:00:00              61.0               0.112903
1         7  2018-05-01 00:00:00              28.0               0.827586
2         8  2021-04-01 00:00:00              51.0               0.063492
3         9  2017-04-01 00:00:00              14.0               0.000000
4        10  2017-05-01 00:00:00              15.0               0.062500

In [52]:
bd_comp_merged.rename(columns={"num_caso": "_id_"},inplace=True)

In [53]:
bd_comp_merged.head()

_id_      mes_informacion  Tiempo_aparicion  mean_comportamiento_j
0     1  2021-05-01 00:00:00              61.0               0.112903
1     7  2018-05-01 00:00:00              28.0               0.827586
2     8  2021-04-01 00:00:00              51.0               0.063492
3     9  2017-04-01 00:00:00              14.0               0.000000
4    10  2017-05-01 00:00:00              15.0               0.062500

In [54]:
bd_final = bd_final.merge(bd_comp_merged, how = "left", on = "_id_")

In [55]:
bd_final.head()

_id_  _id2_  comp_mixto1 fecha_inicio  comp_interno1 comp_interno2  \
0  2333  53659      0.32856   2016-01-04    6079.027356          b'1'   
1   488  11224      0.38498   2016-01-04     911.854103          b'0'   
2  1964  45172      0.34780   2016-01-04       0.000000          b'1'   
3  1220  28060      0.28422   2016-01-04    2482.269504          b'0'   
4  1226  28198      0.31500   2016-01-04       0.000000          b'1'   

   comp_interno3  comp_interno4  categoria1  categoria2  categoria3  \
0           30.0           50.0         2.0         9.0       175.0   
1           19.0           48.0         3.0         3.0       164.0   
2           19.0           53.0         4.0         3.0       164.0   
3            9.0           70.0         3.0         9.0       154.0   
4            3.0           75.0         3.0         9.0       148.0   

   inversion      mes_informacion  Tiempo_aparicion  mean_comportamiento_j  
0        NaN  2021-05-01 00:00:00              52.0               0.076923  
1        NaN  2021-05-01 00:00:00              64.0               0.123077  
2        NaN  2021-05-01 00:00:00              64.0               0.076923  
3        NaN  2021-05-01 00:00:00              58.0               0.030769  
4   0.311552  2018-10-01 00:00:00              33.0               0.029412

In [56]:
bd_final.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 50000 entries, 0 to 49999
Data columns (total 15 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   _id_                   50000 non-null  int64  
 1   _id2_                  50000 non-null  int64  
 2   comp_mixto1            50000 non-null  float64
 3   fecha_inicio           50000 non-null  object 
 4   comp_interno1          50000 non-null  float64
 5   comp_interno2          50000 non-null  object 
 6   comp_interno3          50000 non-null  float64
 7   comp_interno4          50000 non-null  float64
 8   categoria1             50000 non-null  float64
 9   categoria2             50000 non-null  float64
 10  categoria3             50000 non-null  float64
 11  inversion              17536 non-null  float64
 12  mes_informacion        49352 non-null  object 
 13  Tiempo_aparicion       49352 non-null  float64
 14  mean_comportamiento_j  49352 non-null  float64
dtypes:

In [57]:
bd_final.nunique()

_id_                     50000
_id2_                    50000
comp_mixto1               7098
fecha_inicio              1557
comp_interno1              680
comp_interno2                2
comp_interno3               32
comp_interno4               55
categoria1                   6
categoria2                  13
categoria3                  32
inversion                17385
mes_informacion             62
Tiempo_aparicion            65
mean_comportamiento_j     1715
dtype: int64

Consideremos el VaR al 95% para ver a qué clientes les damos crédito según su comportamiento_j promedio, a su vez tomemos el Expected Shortafall al mismo nivel. 

In [58]:
VaR_95 = bd_final["mean_comportamiento_j"].quantile(0.95)
ES_95 = bd_final["mean_comportamiento_j"][bd_final["mean_comportamiento_j"]>=VaR_95].mean()

Consideremos el VaR al 70% para las inversiones, y así considerar a los clientes de mayor inversión

In [77]:
VaR_70 = bd_final["inversion"].quantile(0.7)

In [72]:
bd_final["inversion"].fillna(0, inplace=True)

In [74]:
bd_final.head()

_id_  _id2_  comp_mixto1 fecha_inicio  comp_interno1 comp_interno2  \
0  2333  53659      0.32856   2016-01-04    6079.027356          b'1'   
1   488  11224      0.38498   2016-01-04     911.854103          b'0'   
2  1964  45172      0.34780   2016-01-04       0.000000          b'1'   
3  1220  28060      0.28422   2016-01-04    2482.269504          b'0'   
4  1226  28198      0.31500   2016-01-04       0.000000          b'1'   

   comp_interno3  comp_interno4  categoria1  categoria2  categoria3  \
0           30.0           50.0         2.0         9.0       175.0   
1           19.0           48.0         3.0         3.0       164.0   
2           19.0           53.0         4.0         3.0       164.0   
3            9.0           70.0         3.0         9.0       154.0   
4            3.0           75.0         3.0         9.0       148.0   

   inversion      mes_informacion  Tiempo_aparicion  mean_comportamiento_j  
0   0.000000  2021-05-01 00:00:00              52.0               0.076923  
1   0.000000  2021-05-01 00:00:00              64.0               0.123077  
2   0.000000  2021-05-01 00:00:00              64.0               0.076923  
3   0.000000  2021-05-01 00:00:00              58.0               0.030769  
4   0.311552  2018-10-01 00:00:00              33.0               0.029412

De esta forma consideramos como un cliente viable para darle crédito si este tiene un valor en "mean_comportamiento_j" mayor que el VaR_95 de esta manera reducimos el riesgo de default, en caso de que no lo cumpla, consideramos el caso en que este tenga menor "mean_comportamiento_j" que ES_95 de esta manera hay menor riesgo en casos extremos pero esto lo haremos si es que tiene una buena inversión, es decir, si este tiene un registro de inversión mayor que el VaR_70. En caso de que esto no se cumpla.

In [79]:
l = []
for i in range(len(bd_final)):
    if (bd_final["mean_comportamiento_j"][i] < VaR_95) or (bd_final["mean_comportamiento_j"][i]<ES_95 and bd_final["inversion"][i]>VaR_70):
        l.append(1)
    else:
        l.append(0)

True

In [80]:
bd_final["credito"] = l

In [83]:
bd_final.tail()

_id_    _id2_  comp_mixto1 fecha_inicio  comp_interno1 comp_interno2  \
49995  98879  2274217      0.35158   2021-04-29       0.000000          b'0'   
49996  98990  2276770      0.30434   2021-04-29       0.000000          b'1'   
49997  98669  2269387      0.36922   2021-04-30     607.902736          b'1'   
49998  98830  2273090      0.33334   2021-04-30     709.219858          b'1'   
49999  98810  2272630      0.32780   2021-04-30    1621.073961          b'1'   

       comp_interno3  comp_interno4  categoria1  categoria2  categoria3  \
49995           22.0           74.0         1.0         9.0       167.0   
49996           17.0           60.0         2.0         9.0       162.0   
49997           22.0           60.0         2.0         8.0       167.0   
49998            3.0           53.0         3.0         9.0       148.0   
49999           19.0           74.0         3.0         9.0       164.0   

        inversion      mes_informacion  Tiempo_aparicion  \
49995    0.000000  2021-05-01 00:00:00               1.0   
49996  107.495449  2021-05-01 00:00:00               1.0   
49997    0.000000  2021-05-01 00:00:00               1.0   
49998    0.000000                  NaN               NaN   
49999  388.820490                  NaN               NaN   

       mean_comportamiento_j  credito  
49995                    0.0        1  
49996                    0.0        1  
49997                    0.0        1  
49998                    NaN        0  
49999                    NaN        0

_De esta manera, nuestra base final es bd_final_

In [82]:
bd_final.head(20)

_id_  _id2_  comp_mixto1 fecha_inicio  comp_interno1 comp_interno2  \
0   2333  53659      0.32856   2016-01-04    6079.027356          b'1'   
1    488  11224      0.38498   2016-01-04     911.854103          b'0'   
2   1964  45172      0.34780   2016-01-04       0.000000          b'1'   
3   1220  28060      0.28422   2016-01-04    2482.269504          b'0'   
4   1226  28198      0.31500   2016-01-04       0.000000          b'1'   
5   1278  29394      0.36858   2016-01-04       0.000000          b'1'   
6    899  20677      0.26702   2016-01-04       0.000000          b'1'   
7   2235  51405      0.38000   2016-01-04       0.000000          b'0'   
8    786  18078      0.35078   2016-01-04       0.000000          b'0'   
9   2113  48599      0.37506   2016-01-04       0.000000          b'1'   
10  2410  55430      0.32920   2016-01-04       0.000000          b'0'   
11   277   6371      0.35336   2016-01-04     658.561297          b'1'   
12  2383  54809      0.37110   2016-01-04    2026.342452          b'1'   
13  1468  33764      0.34000   2016-01-04       0.000000          b'1'   
14   924  21252      0.38000   2016-01-04    1519.756839          b'1'   
15    64   1472      0.38000   2016-01-05     709.219858          b'1'   
16  1152  26496      0.33548   2016-01-05       0.000000          b'1'   
17  2421  55683      0.37000   2016-01-05    1114.488349          b'1'   
18  1816  41768      0.37000   2016-01-05    3039.513678          b'1'   
19  1935  44505      0.29150   2016-01-05    1013.171226          b'0'   

    comp_interno3  comp_interno4  categoria1  categoria2  categoria3  \
0            30.0           50.0         2.0         9.0       175.0   
1            19.0           48.0         3.0         3.0       164.0   
2            19.0           53.0         4.0         3.0       164.0   
3             9.0           70.0         3.0         9.0       154.0   
4             3.0           75.0         3.0         9.0       148.0   
5            19.0           45.0         3.0         9.0       164.0   
6             1.0           60.0         1.0         9.0       146.0   
7            15.0           73.0         1.0        11.0       160.0   
8            23.0           68.0         3.0         9.0       168.0   
9            19.0           44.0         3.0         9.0       164.0   
10           15.0           72.0         1.0         5.0       160.0   
11           19.0           57.0         2.0         9.0       164.0   
12           15.0           59.0         1.0         9.0       160.0   
13           19.0           72.0         2.0        11.0       164.0   
14            1.0           53.0         2.0         9.0       146.0   
15           19.0           60.0         1.0         9.0       164.0   
16           14.0           46.0         3.0         9.0       159.0   
17           17.0           52.0         1.0         9.0       162.0   
18           24.0           65.0         1.0         6.0       169.0   
19           21.0           69.0         1.0         9.0       166.0   

     inversion      mes_informacion  Tiempo_aparicion  mean_comportamiento_j  \
0     0.000000  2021-05-01 00:00:00              52.0               0.076923   
1     0.000000  2021-05-01 00:00:00              64.0               0.123077   
2     0.000000  2021-05-01 00:00:00              64.0               0.076923   
3     0.000000  2021-05-01 00:00:00              58.0               0.030769   
4     0.311552  2018-10-01 00:00:00              33.0               0.029412   
5     0.000000  2018-08-01 00:00:00              31.0               0.062500   
6     0.000000  2021-05-01 00:00:00              63.0               0.062500   
7     0.000000  2020-12-01 00:00:00              58.0               0.084746   
8     0.000000  2018-02-01 00:00:00              24.0               0.000000   
9     0.000000  2021-05-01 00:00:00              64.0               0.046154   
10    0.000000  2021-05-01 00:00:00              63.0   

# Pregunta 2

In [ ]:
bd="data_schools.xlsx"
bd_escuelas=pd.read_excel(bd)
bd_escuelas #Vemos si cargó

## Para n

In [ ]:
#Seleccionamos aleatoriamente n escuelas
cor=np.zeros(3200)
for  i in range(0,3200):
    escuelas_random=bd_escuelas.sample(n=15,replace=True)
    cor[i]= spearmanr(escuelas_random.LSAT, escuelas_random.GPA)[0]
print(cor)

In [ ]:
#El error estándar bootstrap es:
err_bt=np.std(cor)
err_bt

In [ ]:
#La correlación sin bootstrap es:
cor_data = spearmanr(bd_escuelas.LSAT, bd_escuelas.GPA)[0]
cor_data

In [ ]:
#Hacemos el gráfico 
fig, ax = plt.subplots(figsize=(9,6))
hist(cor,bins=50,alpha=0.5,edgecolor="white",label="Error estándar Boostrap")
axvline(x=cor_data,color='green', linestyle='--',linewidth=3,label="Correlación sin bootstrap$")
axvline(x=np.mean(cor),color='black', linestyle='--',linewidth=3,label="Promedio de correlación con  Boostrap$")

axvline(percentile(cor,2.5),color="blue",linestyle="--",linewidth=2)                                                                                                                                                    
axvline(percentile(cor,97.5),color="blue",linestyle="--",linewidth=2) 
show()
print("Correlación sin Bootsrap:", cor_data)
print("Promedio de correlación de la muestras Bootstrap:", np.mean(cor))
print("Error estándar las muestras Bootstrap:", err_bt)
print("(", percentile(cor,2.5),",",percentile(cor,97.5),")")

In [ ]:
#Podemos notar que conforme el número de simulaciones Boostrap aumenta, el error estándar
#baja.
estimadores_se_bt=np.zeros(3200)
for i in range(len(cor)):
    estimadores_se_bt[i]=np.std(cor[0:(i+1)])
estimadores_se_bt

In [ ]:
min_1=min(estimadores_se_bt[1:])
min_1

Este es el menor error estándar boostrap

In [ ]:
index_1=(np.asarray(estimadores_se_bt==min_1).nonzero())[0][0]
index_1

In [ ]:
estimadores_se_bt[index_1]

Esta es la cantidad de iteraciones para las que el error estándar estimado con boostrap es menor

## Para n-1

In [ ]:
#Seleccionamos aleatoriamente n-1 escuelas
cor_2=np.zeros(3200)
for  i in range(0,3200):
    escuelas_random=bd_escuelas.sample(n=15,replace=True)
    escuelas_random=escuelas_random.sample(n=14,replace=True)
    cor_2[i]= spearmanr(escuelas_random.LSAT, escuelas_random.GPA)[0]
print(cor_2)

In [ ]:
#El error estándar bootstrap es:
err_bt_2=np.std(cor_2)
err_bt_2

In [ ]:
#La correlación sin bootstrap es:
cor_data = spearmanr(bd_escuelas.LSAT, bd_escuelas.GPA)[0]
cor_data

In [ ]:
#Hacemos el gráfico 
fig, ax = plt.subplots(figsize=(9,6))
hist(cor_2,bins=50,alpha=0.5,edgecolor="white",label="Error estándar Boostrap")
axvline(x=cor_data,color='green', linestyle='--',linewidth=3,label="Correlación sin bootstrap$")
axvline(x=np.mean(cor_2),color='black', linestyle='--',linewidth=3,label="Promedio de correlación con  Boostrap$")

axvline(percentile(cor_2,2.5),color="blue",linestyle="--",linewidth=2)                                                                                                                                                    
axvline(percentile(cor_2,97.5),color="blue",linestyle="--",linewidth=2) 
show()
print("Correlación sin Bootsrap:", cor_data)
print("Promedio de correlación de la muestras Bootstrap:", np.mean(cor_2))
print("Error estándar las muestras Bootstrap:", err_bt_2)
print("(", percentile(cor_2,2.5),",",percentile(cor_2,97.5),")")

In [ ]:
#Podemos notar que conforme el número de simulaciones Boostrap aumenta, el error estándar
#baja.
estimadores_se_bt2=np.zeros(3200)
for i in range(len(cor_2)):
    estimadores_se_bt2[i]=np.std(cor_2[0:(i+1)])
estimadores_se_bt2

In [ ]:
min_2=min(estimadores_se_bt2[10:])
min_2

Este es el menor error estándar boostrap

In [ ]:
index_2=(np.asarray(estimadores_se_bt2==min_2).nonzero())[0][0]
index_2

In [ ]:
estimadores_se_bt2[index_2]

Esta es la cantidad de iteraciones para las que el error estándar estimado con boostrap es menor

## Para toda la muestra

In [ ]:
#Seleccionamos aleatoriamente n=82, que es el tamaño de la muestra, escuelas
cor_3=np.zeros(3200)
for  i in range(0,3200):
    escuelas_random=bd_escuelas.sample(n=82,replace=True)
    cor_3[i]= spearmanr(escuelas_random.LSAT, escuelas_random.GPA)[0]
print(cor_3)

In [ ]:
#El error estándar bootstrap es:
err_bt_3=np.std(cor_3)
err_bt_3

In [ ]:
#La correlación sin bootstrap es:
cor_data = spearmanr(bd_escuelas.LSAT, bd_escuelas.GPA)[0]
cor_data

In [ ]:
#Hacemos el gráfico 
fig, ax = plt.subplots(figsize=(9,6))
hist(cor_3,bins=50,alpha=0.5,edgecolor="white",label="Error estándar Boostrap")
axvline(x=cor_data,color='green', linestyle='--',linewidth=3,label="Correlación sin bootstrap$")
axvline(x=np.mean(cor_3),color='black', linestyle='--',linewidth=3,label="Promedio de correlación con  Boostrap$")

axvline(percentile(cor_3,2.5),color="blue",linestyle="--",linewidth=2)                                                                                                                                                    
axvline(percentile(cor_3,97.5),color="blue",linestyle="--",linewidth=2) 
show()
print("Correlación sin Bootsrap:", cor_data)
print("Promedio de correlación de la muestras Bootstrap:", np.mean(cor_3))
print("Error estándar las muestras Bootstrap:", err_bt_3)
print("(", percentile(cor_3,2.5),",",percentile(cor_3,97.5),")")

In [ ]:
#Podemos notar que conforme el número de simulaciones Boostrap aumenta, el error estándar
#baja.
estimadores_se_bt3=np.zeros(3200)
for i in range(len(cor_3)):
    estimadores_se_bt3[i]=np.std(cor_3[0:(i+1)])
estimadores_se_bt3

In [ ]:
min_3=min(estimadores_se_bt3[20:])
min_3

Este es el menor error estándar boostrap

In [ ]:
index_3=(np.asarray(estimadores_se_bt3==min_3).nonzero())[0][0]
index_3

In [ ]:
estimadores_se_bt3[index_3]

Esta es la cantidad de iteraciones para las que el error estándar estimado con boostrap es menor

# Pregunta 3

In [ ]:
bd2="lineal_horm.xlsx"
bd_horm=pd.read_excel(bd2)
bd_horm

In [ ]:
reg = linear_model.LinearRegression()

In [ ]:
#Vemos el error de la base completa
X = bd_horm[["horas"]]
y = bd_horm[["monto"]]
reg.fit(X,y)
Y_pred = reg.predict(X)
y_estim=Y_pred.reshape(-1)
x=X.to_numpy().reshape(-1)
y_real=y.to_numpy()
y_real=y_real.reshape(-1)
n=len(y_real)
residual=y_real-y_estim
den=sum((x-np.mean(x))**2)
num=sum(residual**2)/(n)
err_reg_completa=np.sqrt(num/den)
err_reg_completa

In [ ]:
#Seleccionamos aleatoriamente n=15 muestras
sims=3200
errores=np.zeros(sims)
for  i in range(0,sims):
    horm_random=bd_horm.sample(n=15,replace=True)
    X = horm_random[["horas"]]
    y = horm_random[["monto"]]
    reg.fit(X, y)
    Y_pred = reg.predict(X)
    y_estim=Y_pred.reshape(-1)
    x=X.to_numpy().reshape(-1)
    y_real=y.to_numpy()
    y_real=y_real.reshape(-1)
    n=len(y_real)
    residual=y_real-y_estim
    den=sum((x-np.mean(x))**2)
    num=sum(residual**2)/(n)
    err_reg=np.sqrt(num/den)
    errores[i]=err_reg
print(errores)

In [ ]:
#Graficamos
fig, ax = plt.subplots(figsize=(9,6))
hist(errores,bins=50,alpha=0.5,edgecolor="white",label="Error estándar Boostrap")
axvline(x=err_reg_completa,color='green', linestyle='--',linewidth=3,label="Error Estándar de b_1 con regresión$")
axvline(x=np.mean(errores),color='black', linestyle='--',linewidth=3,label="Promedio error estándar de b_1 con  Boostrap$")

axvline(percentile(errores,2.5),color="blue",linestyle="--",linewidth=2)                                                                                                                                                    
axvline(percentile(errores,97.5),color="blue",linestyle="--",linewidth=2) 
show()
print("Error Estándar sin Bootsrap:", err_reg_completa)
print("Promedio de Error Estándar de la muestras Bootstrap:", np.mean(errores))
print("Promedio de Error Estándar de la muestras Bootstrap:", np.mean(errores))
print("(", percentile(errores,2.5),",",percentile(errores,97.5),")")

Podemos observar que esta manera de atacar el problema resulta en un error mayor que el generado en clase a través del remuestreo de los errores estándar de una misma regresión. Cambiemos n

In [ ]:
#Seleccionamos aleatoriamente n=15 muestras
sims_2=3200
errores_2=np.zeros(sims_2)
for  i in range(0,sims_2):
    horm_random=bd_horm.sample(n=len(bd_horm[["horas"]]),replace=True)
    X = horm_random[["horas"]]
    y = horm_random[["monto"]]
    reg.fit(X, y)
    Y_pred = reg.predict(X)
    y_estim=Y_pred.reshape(-1)
    x=X.to_numpy().reshape(-1)
    y_real=y.to_numpy()
    y_real=y_real.reshape(-1)
    n=len(y_real)
    residual=y_real-y_estim
    den=sum((x-np.mean(x))**2)
    num=sum(residual**2)/(n)
    err_reg=np.sqrt(num/den)
    errores_2[i]=err_reg
print(errores_2)

In [ ]:
#Graficamos
fig, ax = plt.subplots(figsize=(9,6))
hist(errores_2,bins=50,alpha=0.5,edgecolor="white",label="Error estándar Boostrap")
axvline(x=err_reg_completa,color='green', linestyle='--',linewidth=3,label="Error Estándar de b_1 con regresión$")
axvline(x=np.mean(errores_2),color='black', linestyle='--',linewidth=3,label="Promedio error estándar de b_1 con  Boostrap$")

axvline(percentile(errores_2,2.5),color="blue",linestyle="--",linewidth=2)                                                                                                                                                    
axvline(percentile(errores_2,97.5),color="blue",linestyle="--",linewidth=2) 
show()
print("Error Estándar sin Bootsrap:", err_reg_completa)
print("Promedio de Error Estándar de la muestras Bootstrap:", np.mean(errores_2))
print("Promedio de Error Estándar de la muestras Bootstrap:", np.mean(errores_2))
print("(", percentile(errores_2,2.5),",",percentile(errores_2,97.5),")")

¡Obtuvimos un error menor! en promedio, que el visto en clase

## Para remuestrar la misma cantidad de elementos de cada lote

In [ ]:
#Seleccionamos aleatoriamente n muestras de cada lote
n=5
sims_3=3200
errores_3=np.zeros(sims_3)
for  i in range(0,sims_3):
    horm_random=bd_horm.groupby(['lote']).sample(n=n,replace=True)
    X = horm_random[["horas"]]
    y = horm_random[["monto"]]
    reg.fit(X, y)
    Y_pred = reg.predict(X)
    y_estim=Y_pred.reshape(-1)
    x=X.to_numpy().reshape(-1)
    y_real=y.to_numpy()
    y_real=y_real.reshape(-1)
    n=len(y_real)
    residual=y_real-y_estim
    den=sum((x-np.mean(x))**2)
    num=sum(residual**2)/(n)
    err_reg=np.sqrt(num/den)
    errores_3[i]=err_reg
print(errores_3)

In [ ]:
#Graficamos
fig, ax = plt.subplots(figsize=(9,6))
hist(errores_3,bins=50,alpha=0.5,edgecolor="white",label="Error estándar Boostrap")
axvline(x=err_reg_completa,color='green', linestyle='--',linewidth=3,label="Error Estándar de b_1 con regresión$")
axvline(x=np.mean(errores_3),color='black', linestyle='--',linewidth=3,label="Promedio error estándar de b_1 con  Boostrap$")

axvline(percentile(errores_3,2.5),color="blue",linestyle="--",linewidth=2)                                                                                                                                                    
axvline(percentile(errores_3,97.5),color="blue",linestyle="--",linewidth=2) 
show()
print("Error Estándar sin Bootsrap:", err_reg_completa)
print("Promedio de Error Estándar de la muestras Bootstrap:", np.mean(errores_3))
print("Promedio de Error Estándar de la muestras Bootstrap:", np.mean(errores_3))
print("(", percentile(errores_3,2.5),",",percentile(errores_3,97.5),")")